crates/evm/evm/src/inspectors/revert_diagnostic.rs

```diff
@@ -0,0 +1,207 @@
+        }
+    }
+
+    /// Derives the revert reason based on the cached data. Should only be called after a revert.
+    fn reason(&self) -> Option<DetailedRevertReason> {
+        if let Some((addr, scheme, _)) = self.non_contract_call {
+            let reason = if is_delegatecall(scheme) {
+                DetailedRevertReason::DelegateCallToNonContract(addr)
+            } else {
+                DetailedRevertReason::CallToNonContract(addr)
+            };
+
+            return Some(reason);
+        }
+
+        if let Some((addr, _)) = self.non_contract_size_check {
+            // unknown schema as the call never took place --> output most generic reason
+            return Some(DetailedRevertReason::CallToNonContract(addr));
+        }
+
+        None
+    }
+
+    /// Injects the revert diagnostic into the debug traces. Should only be called after a revert.
+    fn handle_revert_diagnostic(&self, interp: &mut Interpreter) {
+        if let Some(reason) = self.reason() {
+            interp.control.instruction_result = InstructionResult::Revert;
+            interp.control.next_action = InterpreterAction::Return {
+                result: InterpreterResult {
+                    output: reason.to_string().abi_encode().into(),
+                    gas: interp.control.gas,
+                    result: InstructionResult::Revert,
+                },
+            };
+        }
+    }
+}
+
+impl<CTX, D> Inspector<CTX, EthInterpreter> for RevertDiagnostic
+where
+    D: Database<Error = DatabaseError>,
+    CTX: ContextTr<Db = D>,
+    CTX::Journal: JournalExt,
+{
+    /// Tracks the first call with non-zero calldata that targets a non-contract address. Excludes
+    /// precompiles and test addresses.
+    fn call(&mut self, ctx: &mut CTX, inputs: &mut CallInputs) -> Option<CallOutcome> {
+        let target = self.code_target_address(inputs);
+
+        if IGNORE.contains(&target) || ctx.journal_ref().precompile_addresses().contains(&target) {
+            return None;
+        }
+
+        if let Ok(state) = ctx.journal().code(target) {
+            if state.is_empty() && !inputs.input.is_empty() {
+                self.non_contract_call = Some((target, inputs.scheme, ctx.journal_ref().depth()));
+            }
+        }
+        None
+    }
+
+    /// Handles `REVERT` and `EXTCODESIZE` opcodes for diagnostics.
+    ///
+    /// When a `REVERT` opcode with zero data size occurs:
+    ///  - if `non_contract_call` was set at the current depth, `handle_revert_diagnostic` is
+    ///    called. Otherwise, it is cleared.
+    ///  - if `non_contract_call` was set at the current depth, `handle_revert_diagnostic` is
+    ///    called. Otherwise, it is cleared.
+    ///
+    /// When an `EXTCODESIZE` opcode occurs:
+    ///  - Optimistically caches the target address and current depth in `non_contract_size_check`,
+    ///    pending later validation.
+    fn step(&mut self, interp: &mut Interpreter, ctx: &mut CTX) {
```

Comment:
Please remember that `step` and `step_end` are **INCREDIBLY hot functions**, so any minimal change has a huge effect on execution performance

Please look at cheatcodes inspector on how to outline the if conditions inside of step and step_end; in this case you should do something like this:
```
match opcode {
    op::REVERT => self.handle_revert(...),
    op::EXTCODESIZE => self.handle_extcodesize(...),
    _ => {}
}

#[cold]
fn handle_* ...
```